File: AlgoPersonalStudy/src/backjoon/silver/BJ_15650.java
package backjoon.silver;

import java.io.*;
import java.util.*;

public class BJ_15650 {
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] arr = new int[N];
        boolean[] visited = new boolean[N];

        for(int n=0;n<N;n++) arr[n] = n+1;
        backTracking(arr, visited, 0, M);
    }
    static void backTracking(int[] arr, boolean[] visited, int start, int dep){
        if(dep == 0){
            for(int i=0;i<arr.length;i++){
                if(visited[i]) System.out.print(arr[i]+" ");
            }
            System.out.println();
            return;
        }
        for(int i=start;i<arr.length;i++) {
            visited[i] = true;
            backTracking(arr, visited, i + 1, dep - 1);
            visited[i] = false;
        }
    }
}

File: AlgoPersonalStudy/src/_kotlin/backjoon/BJ_15650.kt
package _kotlin.backjoon

import java.io.*
import java.util.*

fun main(){
    val bf = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(bf.readLine())
    val N = Integer.parseInt(st.nextToken())
    val M = Integer.parseInt(st.nextToken())

    val arr = IntArray(N) { n -> n+1 }
    val visited = BooleanArray(N)
    backTracking(arr, visited, 0, M)
}
fun backTracking(arr:IntArray, visited:BooleanArray, start:Int, dep:Int){
    if(dep == 0){
        for(i:Int in arr.indices){
            if(visited[i]) print(""+arr[i] + " ")
        }
        println()
        return
    }
    for(i:Int in start until arr.size){
        visited[i] = true
        backTracking(arr, visited, i+1, dep - 1)
        visited[i] = false
    }
}